tests/test_file.py

```diff
@@ -414,6 +416,62 @@ def try_get():
                 url, requests.codes.bad_request)
         try_get()
 
+    @testmode.standalone
+    def test_optimized_file_checkout(self):
+        """
+        Verifies checkout occurs on first get and not on second.
+        :return:
+        """
+        tempdir = tempfile.gettempdir()
+        self._test_optimized_file_checkout(Replica.aws,"s3", self.s3_test_bucket,
+                                           S3Uploader(tempdir, self.s3_test_bucket))
+        self._test_optimized_file_checkout(Replica.gcp, "gs", self.gs_test_bucket,
+                                           GSUploader(tempdir, self.gs_test_bucket))
+
+    def _test_optimized_file_checkout(self, replica: Replica, scheme: str, test_bucket: str, uploader: Uploader):
+        src_key = generate_test_key()
+        src_data = os.urandom(1024)
+        with tempfile.NamedTemporaryFile(delete=True) as fh:
+            fh.write(src_data)
+            fh.flush()
+
+            uploader.checksum_and_upload_file(fh.name, src_key, "text/plain")
+
+        source_url = f"{scheme}://{test_bucket}/{src_key}"
+
+        file_uuid = str(uuid.uuid4())
+        bundle_uuid = str(uuid.uuid4())
+        version = datetime_to_version_format(datetime.datetime.utcnow())
+
+        # should be able to do this twice (i.e., same payload, different UUIDs)
+        self.upload_file(source_url, file_uuid, bundle_uuid=bundle_uuid, version=version)
+        url = str(UrlBuilder()
+                  .set(path="/v1/files/" + file_uuid)
+                  .add_query("replica", replica.name)
+                  .add_query("version", version))
+
+        now = datetime.datetime.now(pytz.utc)
+        # assert 301 redirect on first GET
+        self.assertGetResponse(url, requests.codes.moved, redirect_follow_retries=0)
```

Comment:
This is outside the scope of this PR, but I think the proper status code for the redirect loop is actually 303 See Other, not 301 Moved Permanently.

Reply:
I see. I'll create an issue for this

Reply:
https://app.zenhub.com/workspace/o/humancellatlas/data-store/issues/1517